flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/BufferBuilder.java

```diff
@@ -40,26 +40,33 @@
 
 	private final SettablePositionMarker positionMarker = new SettablePositionMarker();
 
-	private boolean bufferConsumerCreated = false;
-
 	public BufferBuilder(MemorySegment memorySegment, BufferRecycler recycler) {
 		this.memorySegment = checkNotNull(memorySegment);
 		this.recycler = checkNotNull(recycler);
 	}
 
 	/**
-	 * @return created matching instance of {@link BufferConsumer} to this {@link BufferBuilder}. There can exist only
-	 * one {@link BufferConsumer} per each {@link BufferBuilder} and vice versa.
+	 * @return created matching instance of {@link BufferConsumer} to this {@link BufferBuilder}.
 	 */
 	public BufferConsumer createBufferConsumer() {
-		checkState(!bufferConsumerCreated, "There can not exists two BufferConsumer for one BufferBuilder");
-		bufferConsumerCreated = true;
 		return new BufferConsumer(
 			memorySegment,
 			recycler,
 			positionMarker);
 	}
 
+	/**
+	 * @return created matching instance of {@link BufferConsumer} similar with {@link #createBufferConsumer()},
+	 * except for initializing its reader position based on {@link BufferBuilder}'s current writable position .
+	 */
+	public BufferConsumer createPositionedBufferConsumer() {
```

Comment:
I guess you could drop this method and change the contract of `createBufferConsumer` to:
```
/**
  * (...)
  * This method always creates a `BufferConsumer` starting from the current writer offset. Data written to
  * BufferBuilder before creation of BufferConsumer won't be visible for that BufferConsumer
  **/
public BufferConsumer createBufferConsumer() {
		return new BufferConsumer(
			memorySegment,
			recycler,
			positionMarker,
			positionMarker.cachedPosition);
}
```

and this needs to be covered in the `BufferBuilderAndConsumerTest`

Reply:
Yes, I would supplement a unit test for it.

Reply:
I found if we unify the `createBufferConsumer` as above, it would affect the previous tests which creates `BufferConsumer` after the `BufferBuilder` is filled. I need to refactor these tests to confirm create `BufferConsumer` before writing.